sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -478,77 +512,399 @@ static struct codec_info codec_info_list[] = {
+}
+
+static int create_codec_dai_name(struct device *dev,
+				 const struct snd_soc_acpi_link_adr *link,
+				 struct snd_soc_dai_link_component *codec)
+{
+	unsigned int sdw_version, unique_id, mfg_id;
+	unsigned int link_id, part_id, class_id;
+	int i, id;
+
+	for (i = 0; i < link->num_adr; i++) {
+		sdw_version = SDW_VERSION(link->adr[i]);
+		link_id = SDW_DISCO_LINK_ID(link->adr[i]);
+		unique_id = SDW_UNIQUE_ID(link->adr[i]);
+		mfg_id = SDW_MFG_ID(link->adr[i]);
+		part_id = SDW_PART_ID(link->adr[i]);
+		class_id = SDW_CLASS_ID(link->adr[i]);
+
+		if (is_unique_device(link, sdw_version, mfg_id, part_id,
+				     class_id, i))
+			codec[i].name = devm_kasprintf(dev, GFP_KERNEL,
+						       "sdw:%x:%x:%x:%x",
+						       link_id, mfg_id, part_id,
+						       class_id);
+		else
+			codec[i].name = devm_kasprintf(dev, GFP_KERNEL,
+						       "sdw:%x:%x:%x:%x:%x",
+						       link_id, mfg_id, part_id,
+						       class_id, unique_id);
+
+		if (!codec[i].name)
+			return -ENOMEM;
+
+		id = find_codec_info(part_id, NULL);
+		if (id < 0)
+			return id;
+
+		codec[i].dai_name = codec_info_list[id].dai_name;
+	}
+
+	return 0;
+}
+
+static void set_codec_init_func(const struct snd_soc_acpi_link_adr *link,
+				struct snd_soc_dai_link *dai_links,
+				bool playback)
+{
+	unsigned int part_id;
+	int i, id;
+
+	for (i = 0; i < link->num_adr; i++) {
+		part_id = SDW_PART_ID(link->adr[i]);
+		id = find_codec_info(part_id, NULL);
+
+		if (codec_info_list[id].init)
+			codec_info_list[id].init(link, dai_links,
+						 &codec_info_list[id],
+						 playback);
+	}
+}
+
+static int create_sdw_codec_dai(struct device *dev,
+				struct snd_soc_dai_link *dai_links, int id,
+				struct snd_soc_dai_link_component *cpus,
+				const struct snd_soc_acpi_link_adr *link)
+{
+	struct snd_soc_dai_link_component *codec;
+	unsigned int part_id, link_id;
+	int i, j = 0, idx;
+	int ret;
+
+	codec = devm_kcalloc(dev, link->num_adr,
+			     sizeof(struct snd_soc_dai_link_component),
+			     GFP_KERNEL);
+	if (!codec)
+		return -ENOMEM;
+
+	ret = create_codec_dai_name(dev, link, codec);
+	if (ret < 0)
+		return ret;
+
+	part_id = SDW_PART_ID(link->adr[0]);
+	idx = find_codec_info(part_id, NULL);
+	if (idx < 0)
+		return idx;
+
+	link_id = ffs(link->mask) - 1;
+
+	/* playback & capture */
+	for (i = 0; i < 2; i++) {
+		char *name, *cpu_name;
+
+		if (!codec_info_list[idx].direction[i])
+			continue;
+
+		name = devm_kasprintf(dev, GFP_KERNEL,
+				      sdw_stream_name[i], link_id);
+		if (!name)
+			return -ENOMEM;
+
+		cpu_name = devm_kasprintf(dev, GFP_KERNEL,
+					  "SDW%d Pin%d", link_id, j + 2);
+		if (!cpu_name)
+			return -ENOMEM;
+
+		init_dai_link(dai_links, id + j, name, 1 - i, i, cpus, cpu_name,
+			      codec, link->num_adr, NULL, NULL);
+		set_codec_init_func(link, dai_links + id + j, 1 - i);
+		j++;
+	}
+
+	return j;
+}
+
+static inline int port_mask_to_num(int pmask)
+{
+	int num = 0;
+
+	for (num = 0; pmask; pmask >>= 1)
+		num += pmask & 1;
```

Comment:
isn't this trying to reinvent `hweight_long()`

Reply:
thanks, I will refine it